iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#ifndef IOX2_WAITSET_HPP
+#define IOX2_WAITSET_HPP
+
+#include "iox/duration.hpp"
+#include "iox/expected.hpp"
+#include "iox2/file_descriptor.hpp"
+#include "iox2/internal/iceoryx2.hpp"
+#include "iox2/listener.hpp"
+#include "iox2/service_type.hpp"
+#include "iox2/waitset_enums.hpp"
+
+namespace iox2 {
+/// The [`Guard`] is returned by [`WaitSet::attach_deadline()`], [`WaitSet::attach_notification()`]
+/// or [`WaitSet::attach_interval()`]. As soon as it goes out-of-scope it detaches the attachment.
+/// It can also be used to determine the origin of an event in [`WaitSet::run()`] or
+/// [`WaitSet::run_once()`] via [`AttachmentId::event_from()`] or [`AttachmentId::deadline_from()`].
+template <ServiceType S>
+class Guard {
+  public:
+    Guard(Guard&&) noexcept;
+    auto operator=(Guard&& rhs) noexcept -> Guard&;
+    ~Guard();
+
+    Guard(const Guard&) = delete;
+    auto operator=(const Guard&) = delete;
+
+  private:
+    template <ServiceType>
+    friend class WaitSet;
+
+    template <ServiceType>
+    friend class AttachmentId;
+    explicit Guard(iox2_guard_h handle);
+    void drop();
+
+    iox2_guard_h m_handle = nullptr;
+};
+
+/// Represents an attachment to the [`WaitSet`]
+template <ServiceType S>
+class AttachmentId {
+  public:
+    AttachmentId(const AttachmentId& rhs) = delete;
+    auto operator=(const AttachmentId& rhs) -> AttachmentId& = delete;
+
+    AttachmentId(AttachmentId&& rhs) noexcept;
+    auto operator=(AttachmentId&& rhs) noexcept -> AttachmentId&;
+    ~AttachmentId();
+
+    /// Creates an [`AttachmentId`] from a [`Guard`] that was returned via
+    /// [`WaitSet::attach_interval()`], [`WaitSet::attach_notification()`] or
+    /// [`WaitSet::attach_deadline()`].
+    static auto from_guard(const Guard<S>& guard) -> AttachmentId;
+
+    /// Returns true if an event was emitted from a notification or deadline attachment
+    /// corresponding to [`Guard`].
+    auto event_from(const Guard<S>& guard) const -> bool;
```

Comment:
Why not `has_event` ? The name `event_from` sounds like an event will be returned.

Reply:
My idea is, that it reads like this: `event_from(my_guard);` so that it is nearly like a sentence. I could rename it to `is_event_from(my_guard)`.

@orecham what do you think? The same would also go for `deadline_from`

Reply:
For that purpose, I think `has_event_from` is more natural.

```c++
waitset.has_event_from(my_guard);
```

vs

```c++
waitset.is_event_from(my_guard);
```